Add amend operator to delete ambigious licenses

mmarseu opened the pull request:
This PR adds a new operation to the _amend_ command which deletes licenses identified only by name, with no other context that provides more insight about the license's content.

This PR also eliminates a very similar functionality from the existing `ProcessLicense` operation. Herein lies another breaking change: The old feature eliminated any license whose `name` property included the word "unknown", even if that license includes a text or URL property. The new `DeleteAmbigiousLicenses` operation does not.  
Should that feature be desired, a new operation should be added to do this.

Because the new operation is potentially dangerous (as was the previous one) and shouldn't implicitly run whenever users invoke the _amend_ command, a new mechanism has been introduced that makes operations selectable through command-line switches.

This change is mostly backwards-compatible as most pre-existing operations have been marked _default_ operations so they run when the user doesn't specify anything on the command-line just like they always did.  
a notable exception is the `InferCopyright` operation which must also be considered extremely dangerous. While a decision whether to remove it outright or keep it is pending, it has been turned into a opt-in operation.

The fact that individual operations are now exposed on the CLI for manual selection by the user leads to a few new design considerations for operations which have been documented in the module docstring in `cdxev/amend/operations.py`.

italvi commented on cdxev/amend/operations.py:
```diff
@@ -58,12 +123,19 @@ def _add_bom_ref(self, component: dict) -> None:
-    "known unknowns" should be stated, as we can't guarantee completeness,
-    compositions should be marked as 'incomplete' for SBOMs.
-    This operation erases existing compositions and then adds all components as 'incomplete'.
+    Declares all component compositions as 'incomplete'.
```
As you already removed to add `metadata.component` and state in your commit that it fixes #161, did you change your mind about the `unknown`?

italvi reviewed (changes requested):
@mmarseu I did some tests and seems like most of the commands are working as expected, so after doing the changes, we should be able to merge this one (as the comments are mostly regarding grammar/comments)

However, some more remarks:

- Maybe you should rename this MR as it does more than described and is more a rework than just feat.
- I'm fine with "3. Breaking change: InferCopyright no longer runs by default", now it is up to anybody to decide whether they want to do this or not. Maybe a print to stdout with an `INFO` would be useful, like it is already done in `class LicenseNameToId` and `AddLicenseText`, would even add some consistency and then the changes are traceable 😉.
- In "4. Breaking change: Broken up ProcessLicense" you mention that "It no longer adds an empty string as text to licenses where no text has been found. That intentionally makes the SBOM worse than it was before", however if I only have **one** ambiguous licenses in the `licenses`-array, then `--operation delete-ambiguous-licenses` leads to an empty array, i.e. `licenses: []`. Should not we, in this case, remove the key `licenses` completely, as an empty array does not add any (information) value?
Why the "one" is bold: It seems that `--operation delete-ambiguous-licenses` only deletes the first license with `license.name`, e.g. 
```json
"licenses": [
                {"license": {"id": "Apache-2.0"}},
                {"license": {"name": "Some license"}},
                {"licence": {"name": "blub"}}
            ],
```
leads to
```json
"licenses": [
                {"license": {"id": "Apache-2.0"}},
                {"licence": {"name": "blub"}}
            ],
```
This is something you do not have included in your tests, [there](https://github.com/Festo-se/cyclonedx-editor-validator/pull/160/files#diff-49602f2c763ac9ccd11a12917319f1c1517164d22839d76562599ed2385d5995R457) you only have the case one id and one name.
- The reason why we wanted `supplier`, which you now changed with "5. Breaking change: InferSupplier is now more selective", was the current tool landscape: Most of the tools (e.g. DependencyTrack and BlackDuck) do not understand e.g. `publisher`, mostly they either only supported `supplier` or, additionally, `author`.

mmarseu commented on NOTICE:
```diff
@@ -0,0 +1,6 @@
+CycloneDX Editor Validator
+Copyright (c) Festo SE & Co. KG
```
This is a policy for Apache Software Foundation projects. It is not a requirement for NOTICE files of projects licensed under the Apache license, much less projects like ours licensed under a different license which only include a NOTICE file because one of their dependencies requires it.

If you like, we can still add the year. It's just a small burden to maintain, as you say.

italvi replied:
Is not there an action for that? Or maybe we need an own action, which is triggered on the 1st January, updating the year in the `NOTICE`, `LICENSE` and `mkdocs.yml`?

mmarseu commented on tests/auxiliary/licenses/license_name:
> is this a test, whether the file extension is required?

Precisely!

mmarseu commented on tests/test_main.py:
Urgh, tests 🤢 

Would you be okay with it if I included it in #157 ?

italvi replied:
works for me, will focus on the line coverage there 😜

mmarseu commented:
> @mmarseu I did some tests and seems like most of the commands are working as expected, so after doing the changes, we should be able to merge this one (as the comments are mostly regarding grammar/comments)
> 
> However, some more remarks:
> 
> * Maybe you should rename this MR as it does more than described and is more a rework than just feat.

Fair enough.

> * I'm fine with "3. Breaking change: InferCopyright no longer runs by default", now it is up to anybody to decide whether they want to do this or not. Maybe a print to stdout with an `INFO` would be useful, like it is already done in `class LicenseNameToId` and `AddLicenseText`, would even add some consistency and then the changes are traceable 😉.

Done 

> * In "4. Breaking change: Broken up ProcessLicense" you mention that "It no longer adds an empty string as text to licenses where no text has been found. That intentionally makes the SBOM worse than it was before", however if I only have **one** ambiguous licenses in the `licenses`-array, then `--operation delete-ambiguous-licenses` leads to an empty array, i.e. `licenses: []`. Should not we, in this case, remove the key `licenses` completely, as an empty array does not add any (information) value?

We can. I'll add it.

>   Why the "one" is bold: It seems that `--operation delete-ambiguous-licenses` only deletes the first license with `license.name`, e.g.
> 
> ```json
> "licenses": [
>                 {"license": {"id": "Apache-2.0"}},
>                 {"license": {"name": "Some license"}},
>                 {"licence": {"name": "blub"}}
>             ],
> ```
> 
> leads to
> 
> ```json
> "licenses": [
>                 {"license": {"id": "Apache-2.0"}},
>                 {"licence": {"name": "blub"}}
>             ],
> ```
> 
> This is something you do not have included in your tests, [there](https://github.com/Festo-se/cyclonedx-editor-validator/blob/93a3c3c3d04f59583feb9b09972560f3447ef536/tests/test_amend.py#L457) you only have the case one id and one name.

I've added a test for this. Weirdly, it works perfectly fine for me 😕 

> * Do we need a test for the case that `license.name` is similar to an `SPDX-ID` and the `delete-ambiguous-licenses` operation is used? Just to make sure that it maps it to a `license.id` and does not delete it.

I don't think so. The user is free to make sure that they run `license-name-to-id` before `delete-ambiguous-licenses`. If they don't, they might run into this problem. In practice, this shouldn't be very frequent, because `delete-ambiguous-licenses` doesn't run by default. Most people outside Festo wouldn't ever use it. Internally, we need to make sure our pipeline templates invoke operations in the correct order.

> * The reason why we wanted `supplier`, which you now changed with "5. Breaking change: InferSupplier is now more selective", was the current tool landscape: Most of the tools (e.g. DependencyTrack and BlackDuck) do not understand e.g. `publisher`, mostly they either only supported `supplier` or, additionally, `author`.

I understand. I'll revert this particular change.

italvi commented on docs/available_commands.md:
Agree, as mentioned before, I wanted to start with the recreation of the `README.md`, to get the attention of new users. So I guess, that's a task, I would assign to me? 😅

italvi reviewed (changes requested):
@mmarseu please resolve the merge conflicts and comment how to proceed with the year in `NOTICE`. But otherwise I do not have any further comments 😃

mmarseu commented:
> @mmarseu please resolve the merge conflicts and comment how to proceed with the year in `NOTICE`. But otherwise I do not have any further comments 😃

I've added the year and opened #179 to discuss options for automation.
